File: lib/src/storage_access_framework/document_file_column.dart
/// Representation of the available columns of `DocumentsContract.Document.<Column>`
///
/// [Refer to details](https://developer.android.com/reference/android/provider/DocumentsContract.Document)
class DocumentFileColumn {
  final String _id;

  const DocumentFileColumn._(this._id);

  static const _kPrefix = 'DocumentFileColumn';

  /// Equivalent to [COLUMN_DOCUMENT_ID](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_DOCUMENT_ID)
  static const id = DocumentFileColumn._('$_kPrefix.COLUMN_DOCUMENT_ID');

  /// Equivalent to [COLUMN_DISPLAY_NAME](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_DISPLAY_NAME)
  static const displayName =
      DocumentFileColumn._('$_kPrefix.COLUMN_DISPLAY_NAME');

  /// Equivalent to [COLUMN_MIME_TYPE](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_MIME_TYPE)
  static const mimeType = DocumentFileColumn._('$_kPrefix.COLUMN_MIME_TYPE');

  /// Equivalent to [COLUMN_LAST_MODIFIED](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_LAST_MODIFIED)
  static const lastModified =
      DocumentFileColumn._('$_kPrefix.COLUMN_LAST_MODIFIED');

  /// Equivalent to [COLUMN_SIZE](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_SIZE)
  static const size = DocumentFileColumn._('$_kPrefix.COLUMN_SIZE');

  /// Equivalent to [COLUMN_SUMMARY](https://developer.android.com/reference/android/provider/DocumentsContract.Document#COLUMN_SUMMARY)
  static const summary = DocumentFileColumn._('$_kPrefix.COLUMN_SUMMARY');

  @override
  bool operator ==(Object other) {
    return other is DocumentFileColumn && other._id == _id;
  }

  static const values = <DocumentFileColumn>[
    id,
    displayName,
    mimeType,
    lastModified,
    size,
    summary,
  ];

  @override
  int get hashCode => _id.hashCode;

  @override
  String toString() => _id;
}

File: android/src/main/kotlin/io/alexrintt/sharedstorage/storageaccessframework/lib/DocumentFileColumn.kt
package io.alexrintt.sharedstorage.deprecated.lib

import android.database.Cursor
import android.provider.DocumentsContract
import java.lang.NullPointerException

private const val PREFIX = "DocumentFileColumn"

enum class DocumentFileColumnType {
  LONG,
  STRING,
  INT
}


fun getDocumentsContractColumns(): List<String> {
  return listOf(
    DocumentsContract.Document.COLUMN_DOCUMENT_ID,
    DocumentsContract.Document.COLUMN_DISPLAY_NAME,
    DocumentsContract.Document.COLUMN_MIME_TYPE,
    DocumentsContract.Document.COLUMN_SIZE,
    DocumentsContract.Document.COLUMN_SUMMARY,
    DocumentsContract.Document.COLUMN_LAST_MODIFIED,
  )
}

/// `column` must be a constant String from `DocumentsContract.Document.COLUMN*`
fun typeOfColumn(column: String): DocumentFileColumnType? {
  val values = mapOf(
    DocumentsContract.Document.COLUMN_DOCUMENT_ID to DocumentFileColumnType.STRING,
    DocumentsContract.Document.COLUMN_DISPLAY_NAME to DocumentFileColumnType.STRING,
    DocumentsContract.Document.COLUMN_MIME_TYPE to DocumentFileColumnType.STRING,
    DocumentsContract.Document.COLUMN_SIZE to DocumentFileColumnType.LONG,
    DocumentsContract.Document.COLUMN_SUMMARY to DocumentFileColumnType.STRING,
    DocumentsContract.Document.COLUMN_LAST_MODIFIED to DocumentFileColumnType.LONG,
    DocumentsContract.Document.COLUMN_FLAGS to DocumentFileColumnType.INT
  )

  return values[column]
}

fun cursorHandlerOf(type: DocumentFileColumnType): (Cursor, Int) -> Any? {
  when (type) {
    DocumentFileColumnType.LONG -> {
      return { cursor, index ->
        try {
          cursor.getLong(index)
        } catch (e: NullPointerException) {
          null
        }
      }
    }

    DocumentFileColumnType.STRING -> {
      return { cursor, index ->
        try {
          cursor.getString(index)
        } catch (e: NullPointerException) {
          null
        }
      }
    }

    DocumentFileColumnType.INT -> {
      return { cursor, index ->
        try {
          cursor.getInt(index)
        } catch (e: NullPointerException) {
          null
        }
      }
    }
  }
}
